### PR TITLE
Add bandwidth-minimising mesh sorting routine

### PR DESCRIPTION
**Pull request by dengwirda**
This PR adds a mesh sorting routine `mpas_tools.mesh.creation.sort_mesh` that can be used to reindex an MPAS mesh file to bring the cells, edges and vertices into a more cache-local sequence. This undoes the somewhat 'random' ordering introduced by the mesh conversion tools, and may be desirable re: efficiency for downstream dycore + coupler applications.

The reordering is implemented via a reverse cuthill-mckee approach --- minimising the bandwidth of the cell-to-cell adjacency graph and then permuting the edges and vertices to match. Sorting both culled and unculled meshes is supported, as well as meshes containing additional variables, which are reindexed according to their underlying `nCells`, `nEdges` or `nVertices` dimensions.

The mesh sorter should be callable as a cmd-line tool:
```
python sort_mesh.py \
--mesh-file=path+name-to-mpas-file-for-sort \
--sort-file=path+name-to-sorted-output-file
```
and/or inline:
```
from mpas_tools.mesh.creation.sort_mesh import sort_mesh
mesh = xarray.open_dataset("mpas-file-name")
sort_mesh(mesh)
```
Tested on `pm-cpu` using a QU30km MPAS-O standalone config., which successfully spun-up to 90 days after a sort of the culled mesh. Sorting improves the locality of the various remapping arrays (`verticesOnCell`, `edgesOnVertex`, etc, etc) with all showing smooth distributions.

`edgesOnCell=unsorted`:
eoc_unsorted

`edgesOnCell=issorted`:
eoc_issorted

### CODE REVIEW

**Comment by dengwirda** [comment]
Just keeping this in draft for now @xylar --- before going too far I think we should check whether this reindexing will solve the locality issues wrt. the E3SM coupler and I've reached out to the team re: benchmark cases.

**Comment by xylar** [comment]
@dengwirda. beautiful!  Keep me posted.

**Comment by xylar** [comment]
Picasso, by the way, is turning over in his grave.  The unsorted way it clearly more cubist.

File: conda_package/mpas_tools/mesh/creation/sort_mesh.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    vals[mask] = rev[vals[mask] - 1]
+    return vals
+
+
+def sort_mesh(mesh):
+    """
+    SORT-MESH: sort cells, edges and duals in the mesh
+    to improve cache-locality.
+
+    """
```
Comment (xylar) [inline conda_package/mpas_tools/mesh/creation/sort_mesh.py:28]: Here's the docstring format we use in MPAS-Tools:
```suggestion
    """
    Sort cells, edges and duals in the mesh
    to improve cache-locality.

    Parameters
    ----------
    mesh : xarray.Dataset
        A dataset containing an MPAS mesh to sort
        
    Returns
    -------
    mesh : xarray.Dataset
        A dataset containing the sorted MPAS mesh
    """
```

File: conda_package/mpas_tools/mesh/creation/sort_mesh.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
+
+import numpy as np
+import os
+import xarray
+import argparse
+from scipy.sparse import csr_matrix
+from scipy.sparse.csgraph import reverse_cuthill_mckee
+
+
+def sort_fwd(data, fwd):
```
Comment (xylar) [inline conda_package/mpas_tools/mesh/creation/sort_mesh.py:10]: Let's indicate that this is "private":
```suggestion
def _sort_fwd(data, fwd):
```

File: conda_package/mpas_tools/mesh/creation/sort_mesh.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+    vals = data.values
+    vals = vals[fwd - 1]
+    return vals
+
+
+def sort_rev(data, rev):
```
Comment (xylar) [inline conda_package/mpas_tools/mesh/creation/sort_mesh.py:16]:
```suggestion
def _sort_rev(data, rev):
```
Make this private, move it below the "public" functions and add a brief docstring.

File: conda_package/mpas_tools/mesh/creation/sort_mesh.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+        sort_rev(mesh["cellsOnCell"], cell_rev)
+    mesh["cellsOnEdge"][:] = \
+        sort_rev(mesh["cellsOnEdge"], cell_rev)
+    mesh["cellsOnVertex"][:] = \
+        sort_rev(mesh["cellsOnVertex"], cell_rev)
+
+    for var in mesh.keys():
+        dims = mesh.variables[var].dims
+        if ("nCells" in dims):
+            mesh[var][:] = sort_fwd(mesh[var], cell_fwd)
```
Comment (xylar) [inline conda_package/mpas_tools/mesh/creation/sort_mesh.py:59]:
```suggestion
            mesh[var][:] = _sort_fwd(mesh[var], cell_fwd)
```

File: conda_package/mpas_tools/mesh/creation/sort_mesh.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+    mesh["verticesOnCell"][:] = \
+        sort_rev(mesh["verticesOnCell"], dual_rev)
+
+    mesh["verticesOnEdge"][:] = \
+        sort_rev(mesh["verticesOnEdge"], dual_rev)
+
+    for var in mesh.keys():
+        dims = mesh.variables[var].dims
+        if ("nVertices" in dims):
+            mesh[var][:] = sort_fwd(mesh[var], dual_fwd)
```
Comment (xylar) [inline conda_package/mpas_tools/mesh/creation/sort_mesh.py:84]:
```suggestion
            mesh[var][:] = _sort_fwd(mesh[var], dual_fwd)
```

File: conda_package/mpas_tools/mesh/creation/sort_mesh.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+    mesh["edgesOnEdge"][:] = \
+        sort_rev(mesh["edgesOnEdge"], edge_rev)
+
+    mesh["edgesOnVertex"][:] = \
+        sort_rev(mesh["edgesOnVertex"], edge_rev)
+
+    for var in mesh.keys():
+        dims = mesh.variables[var].dims
+        if ("nEdges" in dims):
+            mesh[var][:] = sort_fwd(mesh[var], edge_fwd)
```
Comment (xylar) [inline conda_package/mpas_tools/mesh/creation/sort_mesh.py:112]:
```suggestion
            mesh[var][:] = _sort_fwd(mesh[var], edge_fwd)
```

File: conda_package/mpas_tools/mesh/creation/sort_mesh.py [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+
+    cell_rev = np.zeros(ncells, dtype=np.int32)
+    cell_rev[cell_fwd - 1] = np.arange(ncells) + 1
+
+    mesh["cellsOnCell"][:] = \
+        sort_rev(mesh["cellsOnCell"], cell_rev)
+    mesh["cellsOnEdge"][:] = \
+        sort_rev(mesh["cellsOnEdge"], cell_rev)
+    mesh["cellsOnVertex"][:] = \
+        sort_rev(mesh["cellsOnVertex"], cell_rev)
```
Comment (xylar) [inline conda_package/mpas_tools/mesh/creation/sort_mesh.py:54]:
```suggestion
    mesh["cellsOnCell"][:] = \
        _sort_rev(mesh["cellsOnCell"], cell_rev)
    mesh["cellsOnEdge"][:] = \
        _sort_rev(mesh["cellsOnEdge"], cell_rev)
    mesh["cellsOnVertex"][:] = \
        _sort_rev(mesh["cellsOnVertex"], cell_rev)
```

File: conda_package/mpas_tools/mesh/creation/sort_mesh.py [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+    dual_fwd = dual_fwd[np.sort(imap)]
+
+    dual_rev = np.zeros(nduals, dtype=np.int32)
+    dual_rev[dual_fwd - 1] = np.arange(nduals) + 1
+
+    mesh["verticesOnCell"][:] = \
+        sort_rev(mesh["verticesOnCell"], dual_rev)
+
+    mesh["verticesOnEdge"][:] = \
+        sort_rev(mesh["verticesOnEdge"], dual_rev)
```
Comment (xylar) [inline conda_package/mpas_tools/mesh/creation/sort_mesh.py:79]:
```suggestion
        _sort_rev(mesh["verticesOnEdge"], dual_rev)
```

Comment (xylar) [reply conda_package/mpas_tools/mesh/creation/sort_mesh.py:79]: etc...  There are many.

File: conda_package/mpas_tools/mesh/creation/sort_mesh.py [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+        dims = mesh.variables[var].dims
+        if ("nEdges" in dims):
+            mesh[var][:] = sort_fwd(mesh[var], edge_fwd)
+
+    mesh["indexToEdgeID"][:] = np.arange(nedges) + 1
+
+    return mesh
+
+
+def cell_del2(mesh):
```
[the remark this answers is not in the document]
Comment (xylar) [reply conda_package/mpas_tools/mesh/creation/sort_mesh.py:119]: Also, move this below `main()` along with the other "private" functions.

Alternatively, if you want to keep this "public", it should have a fleshed out docstring like I suggested for `sort_mesh()`.

File: conda_package/mpas_tools/mesh/creation/sort_mesh.py [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+
+        ncells = mesh.dims["nCells"]
+        nedges = np.count_nonzero(cellsOnCell) // 2
+
+        fptr.write("{} {}\n".format(ncells, nedges))
+        for cell in range(ncells):
+            data = cellsOnCell[cell, :]
+            data = data[data > 0]
+            for item in data:
+                fptr.write("{} ".format(item))
```
Comment (xylar) [inline conda_package/mpas_tools/mesh/creation/sort_mesh.py:184]:
```suggestion
                fptr.write(f"{item} ")
```

**Review by xylar — CHANGES_REQUESTED** [review]
@dengwirda, I have some suggestions that are purely aesthetic to do with integrating this code a bit more with the style of `mpas_tools` overall.  None of this has any bearing on functionality.

I'm going to cherry-pick the current commit onto a branch I have for making release candidates for testing in compass.  I'll let you know how that testing goes.

**Comment by xylar** [comment]
One more thing:  It would be good to add the entry point (redundantly, I know) to this file:
https://github.com/MPAS-Dev/MPAS-Tools/blob/master/conda_package/recipe/meta.yaml
along with a test to make sure you can at least get the help message:
https://github.com/MPAS-Dev/MPAS-Tools/blob/master/conda_package/recipe/meta.yaml#L140
Or if you want, you could test it properly like one of these:
https://github.com/MPAS-Dev/MPAS-Tools/blob/master/conda_package/recipe/meta.yaml#L103

File: conda_package/mpas_tools/mesh/creation/sort_mesh.py [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+        val_edge = np.ones(idx_next.size, dtype=np.int8)
+
+        ivec = np.hstack((ivec, idx_self))
+        jvec = np.hstack((jvec, idx_next))
+        xvec = np.hstack((xvec, val_edge))
+
+    return csr_matrix((xvec, (ivec, jvec)))
+
+
+if (__name__ == "__main__"):
```
Comment (xylar) [inline conda_package/mpas_tools/mesh/creation/sort_mesh.py:153]: To function as an entry point, this needs to be a function:
```suggestion
def main():
```
Then, at the bottom of the file, you can have:
``` python
if (__name__ == "__main__"):\
    main()
```

**Comment by xylar** [comment]
I tested things out in https://github.com/MPAS-Dev/compass/pull/657 and everything worked well.  I need to visualize the `edgesOnCell`, etc. to make sure I'm seeing something similar to the plots above but the mechanics seem to work well.

@dengwirda would you like to make the changes I mentioned above?  If so, that would be my preference but I'm also happy to make them if you don't have time.

**Comment by xylar** [comment]
@dengwirda, using Paraview, I looked at all the following fields from the ECwISC mesh I tested earlier:
```
cellsOnCell, edgesOnCell, verticesOnCell, cellsOnVertex, cellsOnEdge, edgesOnVertex, verticesOnEdge, edgesOnEdge
```
and they look great to me.  I won't post a bunch of identical-looking plots here to prove it.

So I think this is good to go as soon as my earlier clean-up comments are addressed.

**Comment by xylar** [comment]
@dengwirda, I'm seeing this error when sorting the RRS6to18 mesh:
```
Traceback (most recent call last):
  File "/gpfs/fs1/home/ac.xylar/chrysalis/mambaforge/envs/compass_rrs6to18/bin/compass", line 33, in <module>
    sys.exit(load_entry_point('compass', 'console_scripts', 'compass')())
  File "/gpfs/fs1/home/ac.xylar/compass/add_rrs6to18/compass/__main__.py", line 63, in main
    commands[args.command]()
  File "/gpfs/fs1/home/ac.xylar/compass/add_rrs6to18/compass/run/serial.py", line 206, in main
    run_single_step(args.step_is_subprocess)
  File "/gpfs/fs1/home/ac.xylar/compass/add_rrs6to18/compass/run/serial.py", line 167, in run_single_step
    _run_test(test_case, available_resources)
  File "/gpfs/fs1/home/ac.xylar/compass/add_rrs6to18/compass/run/serial.py", line 417, in _run_test
    _run_step(test_case, step, test_case.new_step_log_file,
  File "/gpfs/fs1/home/ac.xylar/compass/add_rrs6to18/compass/run/serial.py", line 468, in _run_step
    step.run()
  File "/gpfs/fs1/home/ac.xylar/compass/add_rrs6to18/compass/ocean/mesh/cull.py", line 165, in run
    cull_mesh(with_critical_passages=True, logger=logger,
  File "/gpfs/fs1/home/ac.xylar/compass/add_rrs6to18/compass/ocean/mesh/cull.py", line 241, in cull_mesh
    _cull_mesh_with_logging(
  File "/gpfs/fs1/home/ac.xylar/compass/add_rrs6to18/compass/ocean/mesh/cull.py", line 374, in _cull_mesh_with_logging
    dsCulledMesh = sort_mesh(dsCulledMesh)
  File "/gpfs/fs1/home/ac.xylar/chrysalis/mambaforge/envs/compass_rrs6to18/lib/python3.10/site-packages/mpas_tools/mesh/creation/sort_mesh.py", line 44, in sort_mesh
    cell_fwd = reverse_cuthill_mckee(cell_del2(mesh)) + 1
  File "/gpfs/fs1/home/ac.xylar/chrysalis/mambaforge/envs/compass_rrs6to18/lib/python3.10/site-packages/mpas_tools/mesh/creation/sort_mesh.py", line 131, in cell_del2
    for edge in range(np.max(topolOnCell)):
  File "<__array_function__ internals>", line 200, in amax
  File "/gpfs/fs1/home/ac.xylar/chrysalis/mambaforge/envs/compass_rrs6to18/lib/python3.10/site-packages/numpy/core/fromnumeric.py", line 2820, in amax
    return _wrapreduction(a, np.maximum, 'max', axis, None, out,
  File "/gpfs/fs1/home/ac.xylar/chrysalis/mambaforge/envs/compass_rrs6to18/lib/python3.10/site-packages/numpy/core/fromnumeric.py", line 86, in _wrapreduction
    return ufunc.reduce(obj, axis, dtype, out, **passkwargs)
ValueError: zero-size array to reduction operation maximum which has no identity
```

I haven't had time to debug at all but wanted to see if you have any guesses.

**Comment by xylar** [comment]
@dengwirda, it looks excellent!  Let me just make another test build with both this and the fixes I have been making in #514.  If my testing goes well, I'll approve and merge this.

**Review by xylar — APPROVED** [review]
After using this sorting in https://github.com/MPAS-Dev/compass/pull/657 and re-testing, I uncovered 2 unrelated issues that my testing of this work happened to expose:
https://github.com/MPAS-Dev/MPAS-Tools/pull/518
https://github.com/MPAS-Dev/compass/pull/661
After those fixes, I was able to run successfully with this mesh sorting for QU240, QUwISC240, EC30to60 and ECwISC30to60 meshes all the way through `files_for_e3sm` in compass.

**Pull request merged**